### DISCUSSION TITLE
how can i use no-root user to install clickhouse

### DISCUSSION THREAD

**Discussion by flyly0755** [ClickHouse/ClickHouse#35786] [Q&A] [answered]
Now I need to install clickhouse on a linux machine, but i can't get the root permission.
it seems generally using rpm cmd to install rpm package with root permission,
if i want to continue to use rpm to install, means i need to change the rpm configuration, so normal user can install, right?

another question is: is there other method  to install clickhouse with normal user except the rpm install method?

**Comment by flyly0755** [comment]
now i try install with tar.gz file, but met with some issue showed as below:

$ cmake ..
-- The C compiler identification is GNU 11.2.0
-- The CXX compiler identification is GNU 11.2.0
-- Check for working C compiler: /usr/local/gcc-11.2.0/bin/cc
-- Check for working C compiler: /usr/local/gcc-11.2.0/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/local/gcc-11.2.0/bin/c++
-- Check for working CXX compiler: /usr/local/gcc-11.2.0/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The ASM compiler identification is GNU
-- Found assembler: /usr/local/gcc-11.2.0/bin/cc
c++ (GCC) 11.2.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

CMake Warning at cmake/tools.cmake:20 (message):
  GCC compiler is not officially supported for ClickHouse.  You should
  migrate to clang.
Call Stack (most recent call first):
  CMakeLists.txt:46 (include)


-- Using custom linker by name: gold
-- Using objcopy: /bin/objcopy
-- Using readelf: /bin/readelf
-- Found Git: /data/tools/git/bin/git (found version "2.30.0")
fatal: not a git repository (or any of the parent directories): .git
-- HEAD's commit hash
fatal: not a git repository (or any of the parent directories): .git
CMake Error at CMakeLists.txt:67 (message):
  Submodules are not initialized.  Run

        git submodule update --init --recursive


-- Configuring incomplete, errors occurred!

the machine to install clickhouse is offline, with this situation how to install successfully clickhouse?
sorry about that i am not familiar with cmake :( so I am a little confused why the installed  process is related with git staff?

**Comment by flyly0755** [comment]
and forget to say, the  clickhouse package i used is ClickHouse-22.3.2.2-lts.tar.gz.

**Comment by enowy** [comment]
I tried with ClickHouse-22.3.18.37-lts.

Did you solve this problem? 
If the submodule problem is not resolved, the build will not work. 
The .gitmodule file exists, but config list doesn't have it (git config --list).

**Comment by enowy** [comment]
How about this link?

"Restore git submodules from .gitmodules"
https://stackoverflow.com/questions/11258737/restore-git-submodules-from-gitmodules

**Comment by mdonkers** [comment]
I don't think you want to compile it locally? Seems like you have the wrong tarball as ClickHouse also has Tgz files with just the binary.

There are various alternatives to install without needing root permissions. The simplest ones not needing root are;
- self-managed install (you may want to skip the `sudo ./clickhouse install` step or run without sudo)
- from Tgz archives, which would unpack the runnable binary locally
- run as Docker container
All described here: https://clickhouse.com/docs/en/install

**Comment by l1t1** [comment, marked as answer]
`curl https://clickhouse.com/ | sh`   can download a self decompression version for your machine

**Comment by enowy** [comment]
If you want to directly compile the source, CMAKE is possible only if you download the submodule in advance.
I think it would be better to study from the submodule of GIT, not how to use CMAKE.

https://github.com/ClickHouse/ClickHouse/discussions/35786#discussioncomment-5189553
